File: java/0740-delete-and-earn.java
class Solution {
    public int deleteAndEarn(int[] nums) {
        Map<Integer, Integer> counter = new HashMap<>();
        for (int i = 0; i < nums.length; i++) {
            counter.put(nums[i], counter.getOrDefault(nums[i], 0) + 1);
        }
        List<Integer> numsList = new ArrayList<>(counter.keySet());
        Collections.sort(numsList);

        int earnOne = 0;
        int earnTwo = 0;
        for (int i = 0; i < numsList.size(); i++) {
            int curEarn = numsList.get(i) * counter.get(numsList.get(i));
            if (i > 0 && numsList.get(i) == numsList.get(i - 1) + 1) {
                int temp = earnTwo;
                earnTwo = Math.max(earnOne + curEarn, earnTwo);
                earnOne = temp;
            } else {
                earnOne = earnTwo;
                earnTwo += curEarn;
            }
        }
        return earnTwo;
    }
}

File: cpp/0740-delete-and-earn.cpp
class Solution {
public:
    int deleteAndEarn(vector<int>& nums) {
        int maxnum=0;
        int trans[10001]={0};
        for(int i:nums){
            trans[i]+=i;
            maxnum=max(maxnum,i);
        }
        int dp[maxnum+1];
        dp[0]=trans[0];
        dp[1]=max(trans[1],trans[0]);
        for(int i=2;i<=maxnum;i++){
            dp[i] = max(dp[i - 1], dp[i - 2] +trans[i]);
        }
        return dp[maxnum];
    }
};
